pkg/build/api/validation/validation.go

```diff
@@ -106,30 +106,40 @@ func ValidateBuildRequest(request *buildapi.BuildRequest) fielderrors.Validation
 	allErrs := fielderrors.ValidationErrorList{}
 	allErrs = append(allErrs, validation.ValidateObjectMeta(&request.ObjectMeta, true, oapi.MinimalNameRequirements).Prefix("metadata")...)
 
-	if request.Revision != nil {
-		allErrs = append(allErrs, validateRevision(request.Revision).Prefix("revision")...)
-	}
 	return allErrs
 }
 
 func validateBuildSpec(spec *buildapi.BuildSpec) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
-	hasSourceType := len(spec.Source.Type) != 0
-	switch t := spec.Strategy.Type; {
-	// 'source' is optional for Custom builds
-	case t == buildapi.CustomBuildStrategyType && hasSourceType:
-		allErrs = append(allErrs, validateSource(&spec.Source).Prefix("source")...)
-	case t == buildapi.SourceBuildStrategyType:
-		allErrs = append(allErrs, validateSource(&spec.Source).Prefix("source")...)
-		if spec.Source.Type == buildapi.BuildSourceDockerfile {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("source.type", nil, "may not be type Dockerfile for source builds"))
+
+	if spec.Source.Git == nil && spec.Source.Binary == nil && spec.Source.Dockerfile == nil {
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("source", spec.Source, "must provide a value for at least one of source, binary, or dockerfile"))
+	}
```

Comment:
This `if` is misplaced. If the strategy is Custom, it is ok to have all of the sources nil.
It seems to be uncovered by unit tests, since none broke. We should add one.